plugins/gzip.ts

```diff
@@ -0,0 +1,57 @@
+import { merge } from "../core/utils/object.ts";
+import { Page } from "../core/file.ts";
+import { pageMatches } from "../core/processors.ts";
+import { gzip as compress, Memory } from "../deps/foras.ts";
+
+import type { Extensions } from "../core/utils/path.ts";
+import type Site from "../core/site.ts";
+
+export interface Options {
+  /** The list of extensions this plugin applies to */
+  extensions?: Extensions;
+
+  /**
+   * Level param between 0 and 9 (9 is the smallest but takes the longest to encode)
+   */
+  level?: number;
+}
+
+// Default options
+export const defaults: Options = {
+  extensions: [".html", ".css", ".js", ".mjs", ".svg", ".json", ".xml", ".txt"],
+  level: 6,
+};
+
+/**
+ * A plugin to compress files with gzip
+ */
+export function gzip(userOptions?: Options) {
+  const options = merge(defaults, userOptions);
+
+  return (site: Site) => {
+    site.addEventListener("beforeSave", () => {
```

Comment:
Why are you using event here?
With a processor, you don't need to import the `pageMatches` function.

```js
site.process(options.extensions, (pages) => {
  for (const page of pages) {
    page.content = compress(page.content);
    page.data.url += ".gz";
  }
});
```

Reply:
To ensure the compression happens last. If I were to use a processor instead, given the following config:

```js
site.use(gzip())
site.use(minifyHTML())
```

It would output:

- `.html` files - minified, uncompresed
- `.html.gz` files - unminified, compresed

So the compressed and uncompressed files divert. Additionally some plugins create/add pages (e.g. the sitemap plugin) which might not get compressed with a processor.

If you think that behavior is ok and want to put the responsibility on the user correctly ordering the plugins I can update the PR to use a processor.

Reply:
For now, I prefer to make the order explicit in the _config.ts file because it's more transparent and is aligned with the behavior of other plugins that requires a especific order (base_path before feed, tailwindcss before postcss, to name only a few).
This also allows to place other plugins after gzip if you don't want to compress the pages that these plugins can generate.

If at some point we need to implement a way to "autosort" the plugings to be independent of the order in the _config.ts file, that solution should work for all plugins.

Reply:
All right - I have updated the PR to use a processor.

In this case we should have a look at the plugins `robots`, `sitemap` and `feed`. They use the `beforeSave` event when they could (probably) use a processor instead (plugins `esbuild` and `on_demand` use the event too but there it seems justified). Because otherwise this:

```js
site.use(sitemap())
site.use(gzip())
```

would only output a `sitemap.xml` without a compressed counterpart.

Reply:
Oh, now I see your point much better.

Yeah, `robots`, `sitemap` and `feed` should be changed to processors. But that is a breaking change.
I think maybe a good solution is to implement something like [render order](https://lume.land/docs/core/render-order/) for processors. For example:

```js
site.process([".html"], fn1, 2); // Third
site.process([".html"], fn2, 1); // Second
site.process([".html"], fn3, -1); // First
```

By default is `0` but some plugins can register processors in different order to ensure they are executed after or before others.

Reply:
That would be one solution. But I think using regular processors and having the user decide on the order is still better. It would be a breaking change to update those plugins but there is no hurry to update them so it can probably wait a bit.

Using something like the render order for processors adds unnecessary complexity imo. Render order works for pages, because you have an overview about _all_ the pages so you can adjust the render order accordingly. With plugins however you dont have an overview about all plugins and their processor order. So some plugin processor orders will still be equal and require ordering from the user and others wont. Some might not even work properly anymore because a plugin needs to run before plugin A with priority 1 and after plugin B with priority 2. This would mean that maintainers of independent plugins might need to talk to other plugin maintainers to find the "right" order that works for that scenario.

Reply:
that's a fair point. 
Okay, let's keep it as is.
Thanks again1